torch/jit/__init__.py

```diff
@@ -134,12 +156,9 @@ def _compile(arg):
             # user passed in), this problem goes away, because the class
             # __init__ is a part of is indeed Foo.
 
-            # Make a copy of the class, with the extra _CompiledMixin base
-            cls = type(arg.__name__, (_CompiledMixin,) + arg.__bases__, dict(arg.__dict__))
-
-            # Monkey-patch forward and __init__ with the compiler versions
-            cls.init_compiler(**kwargs)
-            return cls
+            return type(arg.__name__,
+                        (torch._six.with_metaclass(CompiledModuleMeta, *arg.__bases__),),
```

Comment:
Can you explain this black magic to me?

Reply:
So in Python, it's not like the class instantiates its objects by itself. Classes are regular instances themselves, and their classes are called _metaclasses_. So, just like the call operator for a regular object is implemented in its class, the call operator for a class (that you use to create objects) is implemented in its metaclass. Now, most objects have `type` as their metaclass, and the whole chain of calling `__new__`, then `__init__`, etc. is actually part of `__call__` method of `type`.

I use a metaclass in here, because this allows me to override `__call__`, so I get to do all these modifications right before and after a call to `type.__call__` that actually allocates and initializes the newly created object.

---

As a side note, you can also override `__init__` or `__new__` in your metaclass, and have these methods called every time someone subclasses your base class. We use this mechanism to do some magic in autograd when `torch.autograd.Function` is subclassed to create a custom op (e.g. that's when we create the `FnBackward` object that's not explicitly declared).

Reply:
And the call to `_six` is necessary because Py2 and Py3 require specifying metaclasses in two, completely incompatible ways, so we need certain hacks 😊 (I think Py3's way is even a syntax error in Py2)